Source/WebGPU/WGSL/Overload.cpp

```diff
@@ -0,0 +1,534 @@
+/*
+ * Copyright (c) 2023 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "Overload.h"
+
+#include <wtf/DataLog.h>
+
+namespace WGSL {
+
+static constexpr bool shouldDumpOverloadDebugInformation = false;
+static constexpr ASCIILiteral logPrefix = "> overload: "_s;
+
+template<typename... Arguments>
+inline void log(Arguments&&... arguments)
+{
+    if (shouldDumpOverloadDebugInformation)
+        dataLog(logPrefix, std::forward<Arguments>(arguments)...);
+}
+
+template<typename... Arguments>
+inline void logLn(Arguments&&... arguments)
+{
+    if (shouldDumpOverloadDebugInformation)
+        dataLogLn(logPrefix, std::forward<Arguments>(arguments)...);
+}
+
+struct ViableOverload {
+    const OverloadCandidate* candidate;
+    FixedVector<unsigned> cost;
+    Type* result;
+};
+
+class OverloadResolver {
+public:
+    OverloadResolver(TypeStore&, const WTF::Vector<OverloadCandidate>&, const WTF::Vector<Type*>&, unsigned numberOfTypeSubstitutions, unsigned numberOfValueSubstitutions);
+
+    Type* resolve();
+
+private:
+    static constexpr unsigned s_noConversion = std::numeric_limits<unsigned>::max();
+
+    FixedVector<std::optional<ViableOverload>> considerCandidates();
+    std::optional<ViableOverload> considerCandidate(const OverloadCandidate&);
+    unsigned calculateCost(const AbstractType&, Type*);
```

Comment:
The way overload resolution works in WGSL doesn't have a cost function.

Firstly, only built-in functions are allowed to overload. User-written functions can't overload.

Secondly, all the built-in functions which are overloaded have mutually-exclusive signatures. There's never any situation where 2 function declarations are valid for a single call.

If you have something like `sin(3.0)`, the literal 3.0 gets concretized according to concretization rules, which are not sensitive to the context that it's being passed to a function. So there's only a single correct overload after the concretization.

Reply:
That's interesting. I tried to match the spec as closely as possible as described here: https://www.w3.org/TR/WGSL/#overload-resolution-section

> Firstly, only built-in functions are allowed to overload.

This function doesn't assume anything about the overloads being user defined or builtins, but I don't see how assuming they are builtins wouldn't make the resolution any simpler.

> Secondly, all the built-in functions which are overloaded have mutually-exclusive signatures. There's never any situation where 2 function declarations are valid for a single call.

That's true, but that doesn't mean that the same declaration can't yield more than one resolution. i.e. `+ (T, T) -> T` apply to `1 + 1` can resolve `T` to `AbstractInt`, `i32`, `u32`, `AbstractFloat`, `f32` or `f16`, in that order of preference according to the conversion rank table.

> If you have something like sin(3.0), the literal 3.0 gets concretized according to concretization rules, which are not sensitive to the context that it's being passed to a function. So there's only a single correct overload after the concretization.

You know the spec better than me, but that's not how I interpreted it. The way I understood it, overload resolution doesn't concretize types. e.g. `const x = sin(3.0)`, `x` should have type `AbstractFloat`, but if we use `var y = sin(3.0)`, now `y` should have type `f32`, according to the conversion rank function.

Maybe that's some nuance that I'm missing, and it can be implemented in a simpler way than described in the spec.

Reply:
> This function doesn't assume anything about the overloads being user defined or builtins, but I don't see how assuming they are builtins wouldn't make the resolution any simpler.

It's a finite set - every member of every set of overload resolutions is known before we look at the first line of the user program.

> That's interesting. I tried to match the spec as closely as possible as described here: https://www.w3.org/TR/WGSL/#overload-resolution-section

https://gpuweb.github.io/gpuweb/wgsl/#conversion-rank is totally new to me, I had no idea that there was a ranking algorithm. My background knowledge here comes from the fact that the CG wanted to forbid user functions from overriding each other, precisely because (IIRC) they didn't want to write a general ranking algorithm - they wanted to punt on that algorithm until after the initial version of the language. I guess my memory is out-of-date, since it seems pretty clear that a general ranking algorithm exists.

Sorry for the noise, I think your following the spec was the right thing to do.